pkg/provision/providers/qemu/preflight_linux.go

```diff
@@ -0,0 +1,154 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at http://mozilla.org/MPL/2.0/.
+
+package qemu
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"os/exec"
+	"path/filepath"
+	"runtime"
+	"strings"
+
+	"github.com/coreos/go-iptables/iptables"
+	"github.com/hashicorp/go-getter/v2"
+
+	"github.com/siderolabs/talos/pkg/machinery/constants"
+)
+
+func (check *preflightCheckContext) verifyPlatformSpecific(ctx context.Context) error {
+	for _, check := range []func(ctx context.Context) error{
+		check.cniDirectories,
+		check.cniBundle,
+		check.checkIptables,
+		check.swtpmExecutable,
+		check.checkKVM,
+	} {
+		if err := check(ctx); err != nil {
+			return err
+		}
+	}
+
+	return nil
+}
+
+func (check *preflightCheckContext) cniDirectories(ctx context.Context) error {
+	cniDirs := append([]string{}, check.request.Network.CNI.BinPath...)
+	cniDirs = append(cniDirs, check.request.Network.CNI.CacheDir, check.request.Network.CNI.ConfDir)
+
+	for _, cniDir := range cniDirs {
+		st, err := os.Stat(cniDir)
+		if err != nil {
+			if !os.IsNotExist(err) {
+				return fmt.Errorf("error checking CNI directory %q: %w", cniDir, err)
+			}
+
+			fmt.Fprintf(check.options.LogWriter, "creating %q\n", cniDir)
+
+			err = os.MkdirAll(cniDir, 0o777)
```

Comment:
seems weird a pre-flight check is creating directories, I guess we can move this out somewhere else just before creating the actual cluster, not for this PR